cpp/bench/ann/src/common/benchmark.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "cuda_stub.hpp"  // must go first
+
+#include "ann_types.hpp"
+
+#define JSON_DIAGNOSTICS 1
+#include <nlohmann/json.hpp>
+
+#include <memory>
+#include <unordered_map>
+
+#include <dlfcn.h>
+#include <filesystem>
+
+namespace raft::bench::ann {
+
+struct lib_handle {
+  void* handle{nullptr};
+  explicit lib_handle(const std::string& name)
+  {
+    handle = dlopen(name.c_str(), RTLD_LAZY | RTLD_LOCAL);
+    if (handle == nullptr) {
+      auto error_msg = "Failed to load " + name;
+      auto err       = dlerror();
+      if (err != nullptr && err[0] != '\0') { error_msg += ": " + std::string(err); }
+      throw std::runtime_error(error_msg);
+    }
+  }
+  ~lib_handle() noexcept
+  {
+    if (handle != nullptr) { dlclose(handle); }
+  }
+};
+
+auto load_lib(const std::string& algo) -> void*
+{
+  static std::unordered_map<std::string, lib_handle> libs{};
+  auto found = libs.find(algo);
+
+  if (found != libs.end()) { return found->second.handle; }
+  auto lib_name        = "lib" + algo + "_ann_bench.so";
+  std::string lib_path = "";
+  if (std::getenv("CONDA_PREFIX") != nullptr) {
+    auto conda_path = std::string(std::getenv("CONDA_PREFIX")) + "/bin" + "/ann/";
+    if (std::filesystem::exists(conda_path + "ANN_BENCH")) { lib_path = conda_path; }
+  }
+  if (std::getenv("RAFT_HOME") != nullptr) {
+    auto build_path = std::string(std::getenv("RAFT_HOME")) + "/cpp" + "/build/";
+    if (std::filesystem::exists(build_path + "ANN_BENCH")) { lib_path = build_path; }
+  }
+  return libs.emplace(algo, lib_path + lib_name).first->second.handle;
+}
+
+auto get_fun_name(void* addr) -> std::string
+{
+  Dl_info dl_info;
+  if (dladdr(addr, &dl_info) != 0) {
+    if (dl_info.dli_sname != nullptr && dl_info.dli_sname[0] != '\0') {
+      return std::string{dl_info.dli_sname};
+    }
+  }
+  throw std::logic_error("Failed to find out name of the looked up function");
+}
+
+template <typename T>
+auto create_algo(const std::string& algo,
+                 const std::string& distance,
+                 int dim,
+                 const nlohmann::json& conf,
+                 const std::vector<int>& dev_list,
+                 const nlohman::json& index_conf) -> std::unique_ptr<raft::bench::ann::ANN<T>>
```

Comment:
`const nlohman::json& index_conf` is an extra argument that wasn't added to #1661